File: cpp/795_NumberOfSubarraysWithBoundedMaximum.cpp
// https://leetcode.com/problems/number-of-subarrays-with-bounded-maximum/
class Solution {
public:
    int numSubarrayBoundedMax(vector<int> &A, int L, int R) {
        int ans = 0;
        int lastLargeThanRNumIndex = -1;
        int lastInTheRangeNumIndex = -1;
        for (int i = 0; i < A.size(); ++i) {
            if(A[i] > R) {
                lastLargeThanRNumIndex =  i;
            } else if (A[i] < L ) {
                if(lastInTheRangeNumIndex  != -1 && lastInTheRangeNumIndex > lastLargeThanRNumIndex) {
                    ans += lastInTheRangeNumIndex - lastLargeThanRNumIndex;
                }
            } else {
                lastInTheRangeNumIndex = i;
                ans += i - lastLargeThanRNumIndex;
            }
        }
        return ans;
    }
};

File: swift/Array/795_NumberOfSubarraysWithBoundedMaximum.swift
//https://leetcode.com/problems/number-of-subarrays-with-bounded-maximum/
class Solution {
    func numSubarrayBoundedMax(_ A: [Int], _ L: Int, _ R: Int) -> Int {
        var lastLargeThanRNumIndex = -1
        var lastInTheRangeNumIndex = -1
        var ans = 0
        for i in 0..<A.count {
            if A[i] > R {
                lastLargeThanRNumIndex = i
            } else if A[i] < L {
                if lastInTheRangeNumIndex != -1 && lastInTheRangeNumIndex > lastLargeThanRNumIndex {
                    ans +=  lastInTheRangeNumIndex - lastLargeThanRNumIndex
                }
            } else {
                lastInTheRangeNumIndex = i
                ans += i - lastLargeThanRNumIndex
            }
        }
        return ans
    }
}